### ISSUE TITLE
Netty 4.1.0.CR1: DNS resolver accepts invalid FQDN with label length > 63 octet

### ISSUE THREAD

**Issue by jonas-ren** [netty/netty#4935]
According to https://tools.ietf.org/html/rfc1035 (section 2.3.1), the length of a label in domain name should be 63 octets or less. However, Netty DNS resolver accepts invalid domain name with label greater than 63, and actually send out DNS query with the invalid name. 

So the final outcome of such query depends on the underlying resolver. During my tests, I sometimes see FormErr returned. In other cases, Netty resolver actually got unexpected exception, and caused the request to timeout. See the following stack trace:
```
[WARN] (dns-response-handler-0) io.netty.resolver.dns.DnsNameResolver: [id: 0x4b3274a9, /0:0:0:0:0:0:0:0:50879] Unexpected exception: 
io.netty.handler.codec.CorruptedFrameException: name has an out-of-range pointer
	at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeName(DefaultDnsRecordDecoder.java:116) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeQuestion(DefaultDnsRecordDecoder.java:37) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeQuestions(DatagramDnsResponseDecoder.java:102) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:64) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:154) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:154) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:913) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:94) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:510) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:467) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:381) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:353) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742) [netty-all-4.1.jar:4.1.0.CR1]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_74]
```

I think the resolver should reject such invalid domain name in the first place, and should not send out any DNS query.

**Comment by jonas-ren** [comment]
Encounter another unexpected exception when sending a DNS query with the domain name being "A" repeated 512 times. Interesting that it's not always reproducible even with the same input.
<code>
04 Mar 2016 17:38:15,638 [WARN] (dns-response-handler-0) io.netty.resolver.dns.DnsNameResolver: [id: 0xb0bb7b97, /0:0:0:0:0:0:0:0:56265] Unexpected exception: 
io.netty.handler.codec.DecoderException: java.lang.IndexOutOfBoundsException: readerIndex(480) + length(65) exceeds writerIndex(541): PooledUnsafeDirectByteBuf(ridx: 480, widx: 541, cap: 4096)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.ChannelHandlerInvokerUtil.invokeChannelReadNow(ChannelHandlerInvokerUtil.java:83) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.DefaultChannelHandlerInvoker.invokeChannelRead(DefaultChannelHandlerInvoker.java:154) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:154) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:913) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:94) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:510) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:467) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:381) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:353) [netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742) [netty-all-4.1.jar:4.1.0.CR1]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_74]
Caused by: java.lang.IndexOutOfBoundsException: readerIndex(480) + length(65) exceeds writerIndex(541): PooledUnsafeDirectByteBuf(ridx: 480, widx: 541, cap: 4096)
	at io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1325) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.buffer.AbstractByteBuf.checkReadableBytes(AbstractByteBuf.java:1319) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.buffer.AbstractByteBuf.skipBytes(AbstractByteBuf.java:869) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeName(DefaultDnsRecordDecoder.java:127) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DefaultDnsRecordDecoder.decodeRecord(DefaultDnsRecordDecoder.java:46) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decodeRecords(DatagramDnsResponseDecoder.java:109) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:67) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.dns.DatagramDnsResponseDecoder.decode(DatagramDnsResponseDecoder.java:33) ~[netty-all-4.1.jar:4.1.0.CR1]
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89) ~[netty-all-4.1.jar:4.1.0.CR1]
	... 11 more
</code>

**Comment by trustin** [comment]
Could be fixed together with #4937

**Comment by normanmaurer** [comment]
Fixed
